File: content_agent/brief.py
"""Brief đầu vào của pipeline (§2 SPEC.md)."""

from __future__ import annotations

import json
import re
import unicodedata
from pathlib import Path
from typing import List

from pydantic import BaseModel, Field

from .config import PLATFORM_HINTS

PLATFORMS = tuple(PLATFORM_HINTS)


class VideoBrief(BaseModel):
    """Chỉ `topic` là bắt buộc, phần còn lại có mặc định."""

    topic: str
    platform: str = "youtube_shorts"
    duration_target_sec: int = 60
    tone: str = "giáo dục, gần gũi, hơi hài hước"
    audience: str = "khán giả phổ thông Việt Nam, 18-35 tuổi"
    language: str = "vi"
    must_include: List[str] = Field(default_factory=list)
    avoid: List[str] = Field(default_factory=list)

    def platform_hint(self) -> str:
        return PLATFORM_HINTS.get(self.platform, PLATFORM_HINTS["youtube_shorts"])

    def slug(self) -> str:
        """Slug ASCII dùng đặt tên thư mục output."""
        text = unicodedata.normalize("NFD", self.topic)
        text = "".join(c for c in text if unicodedata.category(c) != "Mn")
        text = text.replace("đ", "d").replace("Đ", "D").lower()
        text = re.sub(r"[^a-z0-9]+", "-", text).strip("-")
        return (text or "video")[:60]

    def as_prompt_block(self) -> str:
        """Phần brief chèn vào prompt của mọi agent."""
        lines = [
            f"- Chủ đề: {self.topic}",
            f"- Nền tảng: {self.platform} ({self.platform_hint()})",
            f"- Thời lượng mục tiêu: {self.duration_target_sec} giây",
            f"- Tone: {self.tone}",
            f"- Khán giả: {self.audience}",
            f"- Ngôn ngữ đầu ra: {self.language}",
        ]
        if self.must_include:
            lines.append("- Bắt buộc có: " + "; ".join(self.must_include))
        if self.avoid:
            lines.append("- Phải tránh: " + "; ".join(self.avoid))
        return "\n".join(lines)

    @classmethod
    def from_file(cls, path: Path) -> "VideoBrief":
        return cls.model_validate(json.loads(Path(path).read_text(encoding="utf-8")))

File: src/brief.ts
// The pipeline's input brief (SPEC.md §2). Only `topic` is required.

import { PLATFORM_HINTS, PLATFORMS } from "./config";

export { PLATFORMS };

export interface VideoBrief {
  topic: string;
  platform: string;
  duration_target_sec: number;
  tone: string;
  audience: string;
  language: string;
  must_include: string[];
  avoid: string[];
}

export const DEFAULT_TONE = "giáo dục, gần gũi, hơi hài hước";
export const DEFAULT_AUDIENCE = "khán giả phổ thông Việt Nam, 18-35 tuổi";

export function makeBrief(fields: Partial<VideoBrief> & { topic: string }): VideoBrief {
  return {
    topic: fields.topic,
    platform: fields.platform || "youtube_shorts",
    duration_target_sec: fields.duration_target_sec ?? 60,
    tone: fields.tone || DEFAULT_TONE,
    audience: fields.audience || DEFAULT_AUDIENCE,
    language: fields.language || "vi",
    must_include: fields.must_include || [],
    avoid: fields.avoid || [],
  };
}

export function platformHint(brief: VideoBrief): string {
  return PLATFORM_HINTS[brief.platform] || PLATFORM_HINTS["youtube_shorts"];
}

export function briefAsPromptBlock(brief: VideoBrief): string {
  const lines = [
    `- Chủ đề: ${brief.topic}`,
    `- Nền tảng: ${brief.platform} (${platformHint(brief)})`,
    `- Thời lượng mục tiêu: ${brief.duration_target_sec} giây`,
    `- Tone: ${brief.tone}`,
    `- Khán giả: ${brief.audience}`,
    `- Ngôn ngữ đầu ra: ${brief.language}`,
  ];
  if (brief.must_include.length) lines.push("- Bắt buộc có: " + brief.must_include.join("; "));
  if (brief.avoid.length) lines.push("- Phải tránh: " + brief.avoid.join("; "));
  return lines.join("\n");
}
